sparse/rest/handlers.go

```diff
@@ -0,0 +1,232 @@
+
+	if index > 0 && remoteHoleInterval.End <= server.holeIntervals[index-1].End {
+		log.Debugf("remoteHoleInterval %s is within server.holeIntervals", remoteHoleInterval)
+	} else {
+		log.Debugf("remoteHoleInterval %s is not within server.holeIntervals", remoteHoleInterval)
+		log.Debug("punching hole:", remoteHoleInterval)
+
+		fiemap := sparse.NewFiemapFile(server.fileIo.GetFile())
+		err := fiemap.PunchHole(remoteHoleInterval.Begin, remoteHoleInterval.Len())
+		if err != nil {
+			errorMessage := fmt.Sprintf("PunchHole: %s error: %s", remoteHoleInterval, err)
+			server.sendErrorResponse(errorMessage, writer)
+			return
+		}
+	}
+}
+
+func (server *SyncServer) getChecksum(writer http.ResponseWriter, request *http.Request) {
+	log.Debug("getChecksum called")
+
+	remoteDataInterval, err := server.getQueryInterval(request)
+	if err != nil {
+		errorMessage := fmt.Sprintf("server.getQueryInterval failed, err: %s", err)
+		server.sendErrorResponse(errorMessage, writer)
+		return
+	}
+	log.Debug("receiving remote interval: ", remoteDataInterval)
+
+	var checksum []byte
+
+	// Search uses binary search to find and return the smallest index i in server.dataIntervals[0, n),
+	// at which closure function f(i) is true. If there is no such index, Search returns n.
+	// n = len(server.dataIntervals)
+	index := sort.Search(len(server.dataIntervals),
+		func(i int) bool { return remoteDataInterval.Begin < server.dataIntervals[i].Begin })
+	log.Debug("found position to insert position in server.dataIntervals is: ", index)
+
+	if index > 0 && remoteDataInterval.End <= server.dataIntervals[index-1].End {
+		log.Debugf("remoteDataInterval %s is within server.dataIntervals", remoteDataInterval)
+		checksum, err = sparse.HashDataInterval(server.fileIo, remoteDataInterval)
+		if err != nil {
+			errorMessage := fmt.Sprintf("HashDataInterval locally: %s failed, err: %s", remoteDataInterval, err)
+			server.sendErrorResponse(errorMessage, writer)
+			return
+		}
+	} else {
+		log.Debugf("remoteDataInterval %s is not within server.dataIntervals, send zero length checksum", remoteDataInterval)
+	}
+	log.Debugf("local checksum length: %d", len(checksum))
+
+	fmt.Fprint(writer, checksum)
+}
+
+func (server *SyncServer) writeData(writer http.ResponseWriter, request *http.Request) {
+	log.Debug("writeData called")
+
+	remoteDataInterval, err := server.getQueryInterval(request)
+	if err != nil {
+		errorMessage := fmt.Sprintf("server.getQueryInterval failed, err: %s", err)
+		server.sendErrorResponse(errorMessage, writer)
+		return
+	}
+	log.Debug("receiving remote interval: ", remoteDataInterval)
+
+	data, err := ioutil.ReadAll(io.LimitReader(request.Body, remoteDataInterval.End-remoteDataInterval.Begin))
+	if err != nil {
+		errorMessage := fmt.Sprintf("ioutil.ReadAll(io.LimitReader(r.Body, end-begin)) error: %s", err)
+		server.sendErrorResponse(errorMessage, writer)
+		return
+	}
+	err = request.Body.Close()
```

Comment:
I don't think you need to call request.Body.Close(), which should be already taken care of.